### ISSUE TITLE
ROS packaging

### ISSUE THREAD

**Issue by chaoflow** [NixOS/nixpkgs#11713]
I'm looking into packaging ros, but before doing so would like some suggestions what you consider the best way to do it.

http://wiki.ros.org/ROS/Installation

Any ideas?

**Comment by sephalon** [comment]
I need ROS packaged too and would be willing to help. However as I'm quite new to NixOS I'm unable to give any advice on the best procedure :-(

**Comment by therealpxc** [comment]
I don't know about all of ROS, but I'm currently packaging some other OSRF items for NixOS. I've got working packages for Gazebo, SDFormat, Ignition-Math2, Ignition-Transport 0.9.0 and Ignition-Transport 1.0.1. I've also got an updated Bullet package which is compatible with Gazebo. Right now I'm not sure if I want to say the packages are complete, so they're just sitting around in my nixos-unstable fork.

Shall we start with those? I can make a new branch based on nixpkgs master or nixos master for the PR.

**Comment by therealpxc** [comment]
I started work today on using Nix to set up the build environment for PX4, as well. Do you guys want to get together online and discuss making a little Nix robotics+UAS maintainers' team?

**Comment by therealpxc** [comment]
The guy who maintains the Arch packages for ROS in the Arch Users' Repository uses this script to generate the PKGBUILDs from the catkin packages. Maybe we could reuse some of [his work](https://github.com/bchretien/arch-ros-stacks) to do a similar thing for Nix?

It also seems like getting just catkin packaged would be a good first step, as would just trying to reproduce in NixOS what they do on Debian with their 'Debian Installation from Source' instructions. This would make sure all of the peripheral dependencies were packaged, and get their build system into nixpkgs.

**Comment by expipiplus1** [comment]
I started adding some python packages in #14798 

I've bumped into some issues packaging `catkin`.

I'm probably going to just install everything in a VM for now, if someone else wants to bring that PR forward I'd be happy to help.

**Comment by acowley** [comment]
I have ROS installed via Nix with one derivation for a distro variant, and I'm now working on atomizing it so we get derivations for each package. I'll open a PR for nixpkgs when it's working, but I'm happy to hear if anyone has thoughts on organization. My current aim is to have, eg, `ros.indigo.perception` as a package set.

**Comment by jyp** [comment]
@acowley : was there any progress? Would you consider sharing your work in its present state so I could build upon it?

**Comment by acowley** [comment]
@jyp: I'll push things today. I updated to kinetic, and added a mechanism to extend a ROS variant package set (e.g. `perception`) with custom packages. I don't yet know how best to organize things, so I'll have to write up a quick tour guide, too.

**Comment by acowley** [comment]
@jyp [Here it is](https://github.com/acowley/ros2nix)

I'd like to roll-up the use of the actual `ros2nix` executable in the `ros.nix` file, but haven't done so yet so it is a two-step process. Also, do bear in mind that building everything takes a long time.

**Comment by jyp** [comment]
Awesome, thanks. We'll be deploying ROS in the next few weeks, I'll let you
know how it goes.

**Comment by akru** [comment]
Based on @acowley great job I make ROS Lunar `ros_comm` package set channel:
https://hydra.aira.life/project/aira/channel/latest
Also x86_64 binary cache available at https://hydra.aira.life

**Comment by awesomebytes** [comment]
Hey guys, I'm interested in having ROS running (and specially, automatically compiled) by using nix. May I know what's your current status? (I'm new to nix)

**Comment by therealpxc** [comment]
I'm no longer working at a robotics company, so I haven't touched related packages in a long time. But I would like to at least update Gazebo one last time, and to record my lapsed maintainership by removing myself from that list.

I'm also happy to work with anyone (@acowley ?) who is still working with ROS stuff to get a solution into Nixpkgs.

@akru it looks like you're still building those packages on your Hydra instance. Are you still using them, as well?

**Comment by akru** [comment]
@therealpxc Yes, currently I have minimalistic port of ROS packages - ros_comm package stack. It builded on my hydra and used in my company for supporting ROS in NixOS. I’ll really happy if anyone helps me to integrate more complex ROS packages into nixpkgs tree.

**Comment by nuclearsandwich** [comment]
Howdy, I work on release tooling for the ROS project. We've been working on some tooling that might make it easier to support packaging ROS for Nix. [superflore](github.com/ros-infrastructure/superflore) is a tool for ingesting ROS distribution data and generating packaging infrastructure for different systems. It's very much a new project and currently supports gentoo and openembedded but if it looks useful we'd gladly accept contributions to add support for Nix/NixOS.

**Comment by acowley** [comment]
That effort is great to see, @nuclearsandwich! Is there anything yet written about how the pieces fit together for supporting a different packaging system?

Something I'll note about my efforts is that I ran into some pretty unusual corner cases. The approach I was taking was to break everything up into distinct nix derivations, then build a development environment that pulls them all in. As it happens, I ran into problems with my shell environment getting too big due to separate, long paths being piled into `PYTHONPATH` and `ROS_PACKAGE_PATH`. I ended up needing to write custom wrappers for `clang` on macOS to make better use of response files as the command lines were too long, and I had to script some pruning of the environment before even calling `exec` as it too was groaning under the thousands of paths. Even so, building with `catkin` is still very slow (much slower than on a normal Linux setup), I think due in large part to these things. Other minor issues arose, too, but I don't think my circumventions were interesting as they were a consequence of the epic search path design problem.

What I'd like to try is more carefully producing a nix store path that collects symlinks to everything in the particular ROS distro to make the paths a bit more humane.

**Comment by szobov** [comment]
@akru would it be possible to share your ports of ROS packages?

**Comment by akru** [comment]
@szobov yes, sources [here](https://github.com/airalab/airapkgs/tree/nixos-unstable/pkgs/development/ros-modules), and channel/cache [here](https://hydra.aira.life/project/aira/channel/latest).

**Comment by nuclearsandwich** [comment]
> Is there anything yet written about how the pieces fit together for supporting a different packaging system?

Not yet. Currently there's some duplicated effort between bloom and superflore with respect to generating templates (although they target different platforms). I think integrating them better so that template creation and population for all platforms is done via bloom and invoked via superflore when building whole rosdistros at once is probably higher in my priority queue than documenting the current state. But I'm not focusing on either at present as they're in a bit of a steady state for our current needs.

Regarding the path size challenges: I'm not too familiar with how Nix manages things under the hood. ROS tools build their own scripts for managing overlays of workspaces and if my weak understanding of Nix isn't totally incorrect it uses symlinks to unify isolated packages under a virtual prefix? Does your current strategy use one or the other or both of these?

**Comment by acowley** [comment]
@nuclearsandwich Some languages have tooling in nixpkgs that builds a directory full of symlinks, but a typical Nix thing is to say you want to try out a couple programs without installing them. In that case, you could say, `nix-shell -p ruby imagemagick` and you'd have a shell where the binaries associated with those packages are on `PATH`. That's the approach that doesn't scale to rosdistro variants too far beyond `ros_comm`. ROS somewhat exacerbates the potential problems with the environment variable strategy because you have `PATH`, `ROS_PACKAGE_PATH`, and `PYTHONPATH` off the top of my head.

I don't think this is a fatal issue in general, just an observation I wanted to share because it's not often you have things like `exec` fail due to environment size! 😛

**Comment by nuclearsandwich** [comment]
Thanks for clarifying. If I may summarize: `nix-shell -p .... 30 or 40 ROS packages` is fragile. Does nix have any notion of metapackages or package groups that would enable the use of say `nix-shell -p ros-kinetic` backed by a single or minimal number of path extensions?


> I don't think this is a fatal issue in general, just an observation I wanted to share because it's not often you have things like exec fail due to environment size!

No indeed. It's a neat trick!

**Comment by acowley** [comment]
The helper one should use is [symlinkJoin](https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/trivial-builders.nix#L76). It does the symlink work. In addition to that, one may need additional wrapping to inform the gathered packages of their common home. The python tooling [does that](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/interpreters/python/wrapper.nix#L28) by creating a wrapper script for each package in any ad hoc package set that sets `PYTHONHOME` before calling into the wrapped script.

**Comment by breakds** [comment]
Hi everyone! Not sure if anyone is still working on this.

I am working on a self-driving project and would like to get ROS packaged with nix. I am fairly new to nix packaging, so this is quite ambitious. I see many good progress being made in this thread, and am wondering whether I can join the effort and learn from everyone.

I am targeting ros kinetic. So I have been able to build the python dependencies from http://wiki.ros.org/kinetic/Installation/Source

With that, I started to package each single module from ros. I pick catkin as the first one to attack. @akru 's [catkin package](https://github.com/airalab/airapkgs/blob/nixos-unstable/pkgs/development/ros-modules/catkin/default.nix) is a very good start for me (Thanks!), but I was not able to get it working (tried to patch the cmake files, but still cannot fix it).

In short, I am wondering:

1. Is there anyone who is still working on this? If so, can I join the effort to learn instead of fighting alone?
2. Is my strategy correct (to package each ros-module one by one, fighting with cmake)?
3. I am currently working on it in one of my own repos - should it be eventually moved to nixpkgs? I think as a newbie I might be doing many things wrong as there might be some coding/structure style guide for nix.

Thanks a lot!

**Comment by akru** [comment]
Hi @breakds! Thank you for your interest and welcome to NixOS ROS community!

I’ve try to describe my approach of ROS development under NixOS in this post: https://github.com/airalab/aira/issues/75#issuecomment-440318394

**Comment by breakds** [comment]
Thank you for the prompt reply, @akru ! Thanks for sharing your ROS development approach - looks reasonable to me.

Have some more newbie questions if you don't mind:

1. It seems that pkgs in airalab is a fork from the official nixpkgs. What is the best way for me to try the ROS packages there without breaking my nixos setup which is currently based on the official nixpkgs? So far I have been developing `.nix` files in my own project directory, and run `nix-build` on them manually - does that sound wrong?

2. I tried copy your `catkin.nix` as shown in [this repo](https://github.com/breakds/nixros/blob/master/ros-modules/catkin.nix) with slight modification. Does this seem sane to you (I think I may be making newbie's mistake here). It seems to me that in this version of catkin the generated scripts from cmake (Catkin's CMakeLists.txt is generating installation scripts to install, by inspecting it), where some of the scripts has various assumptions of existed python packages or shebangs. I followed your example to patch them, but seemingly not enough. Is there any other typical patches that I could use to work this around?

Really appreciate your help!

**Comment by akru** [comment]
@breakds Currently for building ROS packages I use approach with custom nixpkgs fetch script:https://github.com/airalab/robonomics_comm/blob/master/fetchNixpkgs.nix and use custom nixpkgs for building package https://github.com/airalab/robonomics_comm/blob/master/release.nix.

Probably overlay approach could be better but for me custom nixpkgs enough. Unfortunately overlays doesn’t support nixos services that we also implement in Airalab. I periodically sync it with official nixpkgs.

**Comment by breakds** [comment]
Thanks for the pointers, @akru ! Also haven't heard of overlays, but I will look it up.

**Comment by lopsided98** [comment]
I have an overlay that provides most ROS packages: https://github.com/lopsided98/nix-ros-overlay. I have been using it for classes and research for over a year now. It is semi-automatically updated and a binary cache is available on Cachix.

**Comment by Renha** [comment]
Cool! I'm still interested in more or less complete packaging

**Comment by dawagner** [comment]
So far, I'm happy with @lopsided98 's overlay :)

**Comment by Pandapip1** [comment]
@lopsided98 would you be okay with me trying to merge https://github.com/lopsided98/nix-ros-overlay into nixpkgs?

**Comment by Pandapip1** [comment]
Self-assigning this as I'm actively working on porting https://github.com/lopsided98/nix-ros-overlay to nixpkgs.

**Comment by lopsided98** [comment]
I don't have any objection to merging the various dependencies into nixpkgs, but it is not something I have time to work on. On the other hand, most of the overlay consists of thousands of auto-generated packages that are probably not appropriate for nixpkgs.

**Comment by Pandapip1** [comment]
> On the other hand, most of the overlay consists of thousands of auto-generated packages that are probably not appropriate for nixpkgs.

Nixpkgs has several auto-generated package sets, most notably `rPackages`.

**Comment by Pandapip1** [comment]
As part of an optional packaging protocol, I am pleased to present an amusing fact. [This is the oldest package request that's still open](https://github.com/NixOS/nixpkgs/issues?page=1&q=is%3Aissue+is%3Aopen+sort%3Acreated-asc)

**Comment by wentasah** [comment]
I agree with @lopsided98 that moving everything to nixpkgs is overkill, at least at this time. My experience is that ROS, especially ROS 2, is full of various bugs and quick iteration in fixing these bugs (preferably upstream) is useful. ROS is released every few weeks and keeping us with that in nixpkgs would be difficult (unless you're a dedicated nixpkgs committer).

That is not to say, that iterating in **nix-ros-overlay** is currently super fast. But that's IMHO caused by lack of reviewers. My experience with @lopsided98 maintenance of the overlay is that if PRs are properly reviewed he merges them pretty quickly. To make the progress even faster, I've setup a [public Hydra instance](https://hydra.iid.ciirc.cvut.cz/project/nix-ros-overlay) that builds the overlay and everybody can see the current status. And I have plans to also add buildbot-nix, which could build PRs even before merging.

Another thing, which may make iterating on ROS with Nix faster is my [ros2nix](https://github.com/wentasah/ros2nix) tool. With it, you can create Nix expressions directly from `package.xml`. I use it, for example, to test whether a development version of a ROS package (not yet present in a ROS release) fixes the problems I have. If not, I can easily create and test a fix, which works with Nix and submit it upstream.

So I'd like to invite everybody interested in ROS to cooperate in [nix-ros-overlay](https://github.com/lopsided98/nix-ros-overlay), where it should be faster than in nixpkgs. After ROS is stable and boring (if ever), we can think about merging that to nixpkgs.
